File: examples/transport_poc/wasm_node_test/transport.test.mjs
import assert from "node:assert/strict";
import { wireOk, wireErr } from "../../../runtime/typescript/src/wire.ts";
import {
  registerTransport,
  runTransportRoundtrip,
} from "../dist/wasm/generated/transport_poc_node.ts";

class FakeTransport {
  #config = null;
  #buffer = [];

  async configure(config) {
    this.#config = config;
    return wireOk(undefined);
  }

  async writeAll(data) {
    if (this.#config === null) {
      return wireErr({ tag: "NotConfigured" });
    }
    this.#buffer.push(...data);
    return wireOk(undefined);
  }

  async read(maximumBytes, _timeout) {
    if (this.#config === null) {
      return wireErr({ tag: "NotConfigured" });
    }
    const take = Math.min(maximumBytes, this.#buffer.length);
    const chunk = Uint8Array.from(this.#buffer.splice(0, take));
    return wireOk(chunk);
  }
}

class FailingTransport {
  async configure(_config) {
    return wireErr({ tag: "Timeout" });
  }
  async writeAll(_data) {
    return wireErr({ tag: "Timeout" });
  }
  async read(_maximumBytes, _timeout) {
    return wireErr({ tag: "Timeout" });
  }
}

async function testRoundTrip() {
  const transport = new FakeTransport();
  const payload = Uint8Array.from([1, 2, 3, 4, 5]);

  const result = await runTransportRoundtrip(transport, payload);

  assert.deepEqual(Array.from(result), Array.from(payload));
  console.log("PASS: round trip through wasm + JS-implemented Transport");
}

async function testErrorPropagation() {
  const transport = new FailingTransport();
  const payload = Uint8Array.from([9]);

  await assert.rejects(() => runTransportRoundtrip(transport, payload));
  console.log("PASS: TransportError from JS propagates back through wasm");
}

// registerTransport/unregisterTransport aren't exercised directly by
// runTransportRoundtrip's own bookkeeping, but touch them to make sure the
// low-level handle registration path itself doesn't throw.
function testHandleRegistration() {
  const handle = registerTransport(new FakeTransport());
  assert.ok(handle !== 0, "handle should be non-zero");
  console.log("PASS: callback handle registration");
}

await testHandleRegistration();
await testRoundTrip();
await testErrorPropagation();

console.log("\nAll wasm/JS Transport tests passed.");

File: examples/transport_poc/dart_test/test/transport_test.dart
import 'dart:typed_data';

import 'package:test/test.dart';
import 'package:transport_poc/transport_poc.dart';

final class FakeTransport implements Transport {
  TransportConfig? _config;
  final List<int> _buffer = [];

  @override
  Future<BoltFFIResult<void, TransportError>> configure(
    TransportConfig config,
  ) async {
    _config = config;
    return BoltFFIResult.ok(null);
  }

  @override
  Future<BoltFFIResult<void, TransportError>> writeAll(
    Uint8List data,
  ) async {
    if (_config == null) {
      return BoltFFIResult.err(TransportError.notConfigured());
    }
    _buffer.addAll(data);
    return BoltFFIResult.ok(null);
  }

  @override
  Future<BoltFFIResult<Uint8List, TransportError>> read(
    int maximumBytes,
    Duration timeout,
  ) async {
    if (_config == null) {
      return BoltFFIResult.err(TransportError.notConfigured());
    }
    final take = maximumBytes < _buffer.length ? maximumBytes : _buffer.length;
    final chunk = Uint8List.fromList(_buffer.take(take).toList());
    _buffer.removeRange(0, take);
    return BoltFFIResult.ok(chunk);
  }
}

final class FailingTransport implements Transport {
  @override
  Future<BoltFFIResult<void, TransportError>> configure(
    TransportConfig config,
  ) async => BoltFFIResult.err(TransportError.timeout());

  @override
  Future<BoltFFIResult<void, TransportError>> writeAll(
    Uint8List data,
  ) async => BoltFFIResult.err(TransportError.timeout());

  @override
  Future<BoltFFIResult<Uint8List, TransportError>> read(
    int maximumBytes,
    Duration timeout,
  ) async => BoltFFIResult.err(TransportError.timeout());
}

void main() {
  test('Dart-implemented Transport round-trips through Rust', () async {
    final transport = FakeTransport();
    final payload = Uint8List.fromList([1, 2, 3, 4, 5]);

    final result = await runTransportRoundtrip(transport, payload).value;

    expect(result, payload);
  });

  test('errors returned from Dart propagate back through Rust', () async {
    final transport = FailingTransport();
    final payload = Uint8List.fromList([9]);

    await expectLater(
      runTransportRoundtrip(transport, payload).value,
      throwsA(isA<TransportError>()),
    );
  });
}
